pilot/pkg/bootstrap/namespacecontroller.go

```diff
@@ -82,44 +114,47 @@ func NewNamespaceController(ca *ca.IstioCA, core corev1.CoreV1Interface) (*Names
 // Run starts the NamespaceController until a value is sent to stopCh.
 func (nc *NamespaceController) Run(stopCh <-chan struct{}) {
 	go nc.namespaceController.Run(stopCh)
+	go nc.configMapController.Run(stopCh)
 }
 
-// When a namespace is created, Citadel adds its public CA certificate
+func (nc *NamespaceController) getData() map[string]string {
+	return map[string]string{
+		constants.CACertNamespaceConfigMapDataName: string(nc.ca.GetCAKeyCertBundle().GetRootCertPem()),
+	}
+}
+
+func (nc *NamespaceController) writeToNamespace(ns string) {
+	meta := metav1.ObjectMeta{
+		Name:      CACertNamespaceConfigMap,
+		Namespace: ns,
+		Labels:    configMapLabel,
+	}
+	err := certutil.InsertDataToConfigMap(nc.core, meta, nc.getData())
+	if err != nil {
```

Comment:
insertDataToConfigMap could fail for transient reasons (connection drop). Since we don't resync we should retry the work at some point in the future. A workqueue could help here (see https://github.com/kubernetes/sample-controller/blob/d90ced32cf661b9821ecb20b08502ad62377a035/controller.go#L150).

Reply:
Yeah that probably makes more sense, and just replace the retries with the queue?

Reply:
Ah, I see you added some retry logic to InsertDataToConfigMap to mitigate this. This could still (rarely) leak updates. IMO a workqueue would be more reliable and is more idiomatic as well.

Reply:
Our comments crossed paths. Ya, replacing retry with a queue should be good. The basic flow is:

1) add/update/delete insert the object key into the workqueue
2) a worker thread pulls items from the queue and processes them
3) If work fails (e.g. insertDataToConfigMap failed) the work item is requeued.